### ISSUE TITLE
CA1508: False positive in cast

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn-analyzers#5140]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

The example below produces a warning, which it should not.

### Steps To Reproduce

```c#
#nullable enable

private static int GetInt32()
{
    object? value = GetNullableObject();
    return (int?)value ?? 0;
    //     ^^^^^^^^^^^ CA1508
}

private static object? GetNullableObject()
{
    return 1;
    //return null;
}
```

### Expected behavior
No warning.

### Actual behavior
Warning: '(int?)value' is never 'null'. Remove or refactor the condition(s) to avoid dead code.
